File: src/games.js
const database = require('./database');
const { transient } = require('./utils/ui');
const { normalizeWord, isTurkishDictionaryWord } = require('./dictionary');

const queues = new Map();

async function isDictionaryWord(word) {
  return isTurkishDictionaryWord(word);
}

async function rejectMessage(message, reason) {
  if (database.getConfig(message.guild.id, 'game_delete_invalid', true)) await message.delete().catch(() => null);
  await transient(message.channel, `❌ <@${message.author.id}> ${reason}`, 5500);
}

async function handleCounting(message) {
  const state = database.getGameState(message.guild.id);
  const expected = state.counting_value + 1;
  if (!/^\d+$/.test(message.content.trim()) || Number(message.content.trim()) !== expected || state.counting_user_id === message.author.id) {
    const sameUser = state.counting_user_id === message.author.id;
    const shouldReset = database.getConfig(message.guild.id, 'counting_reset_on_error', true);
    if (shouldReset) database.resetGame(message.guild.id, 'counting');
    await rejectMessage(message, sameUser
      ? `aynı kişi art arda sayamaz.${shouldReset ? " Oyun **1**'den yeniden başladı." : ` Sıradaki sayı hâlâ **${expected}**.`}`
      : `sıradaki sayı **${expected}** olmalıydı.${shouldReset ? " Oyun **1**'den yeniden başladı." : ''}`);
    return;
  }
  database.setCountingState(message.guild.id, expected, message.author.id);
  await message.react(expected % 100 === 0 ? '💯' : '✅').catch(() => null);
}

async function handleWordChain(message) {
  const word = normalizeWord(message.content);
  const minLength = database.getConfig(message.guild.id, 'word_min_length', 2);
  if (!new RegExp(`^[aâbcçdefgğhıiîjklmnoöprsştuüûvyz]{${minLength},30}$`).test(word)) {
    return rejectMessage(message, `yalnızca ${minLength}–30 harflik tek bir Türkçe kelime yazmalısın.`);
  }

  const state = database.getGameState(message.guild.id);
  if (state.word_user_id === message.author.id) return rejectMessage(message, 'aynı kişi art arda kelime yazamaz.');
  if (state.last_word && word[0] !== state.last_word.at(-1)) {
    return rejectMessage(message, `kelime **${state.last_word.at(-1).toLocaleUpperCase('tr-TR')}** harfiyle başlamalı.`);
  }
  if (database.hasUsedWord(message.guild.id, word)) return rejectMessage(message, `**${word}** daha önce kullanıldı.`);

  const valid = await isDictionaryWord(word);
  if (!valid) return rejectMessage(message, `**${word}** yerel Türkçe oyun sözlüğünde bulunamadı.`);

  database.setWordState(message.guild.id, word, message.author.id);
  await message.react('✅').catch(() => null);
}

function enqueue(guildId, operation) {
  const previous = queues.get(guildId) || Promise.resolve();
  const current = previous.catch(() => null).then(operation).finally(() => {
    if (queues.get(guildId) === current) queues.delete(guildId);
  });
  queues.set(guildId, current);
  return current;
}

async function handleGameMessage(message) {
  const settings = database.getSettings(message.guild.id);
  if (settings.counting_enabled && message.channel.id === settings.counting_channel_id) {
    await enqueue(message.guild.id, () => handleCounting(message));
    return true;
  }
  if (settings.word_chain_enabled && message.channel.id === settings.word_chain_channel_id) {
    await enqueue(message.guild.id, () => handleWordChain(message));
    return true;
  }
  return false;
}

module.exports = { handleGameMessage, normalizeWord, isDictionaryWord };

File: internal/bot/games.go
package bot

import (
	"bufio"
	_ "embed"
	"regexp"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/bwmarrin/discordgo"
)

var wordPattern = regexp.MustCompile(`^[aâbcçdefgğhıiîjklmnoöprsştuüûvyz]{2,30}$`)
var dictionaryOnce sync.Once
var dictionaryWords []string

//go:embed assets/turkish_dictionary.txt
var dictionaryData string

func normalizeWord(s string) string {
	s = strings.TrimSpace(s)
	s = strings.ReplaceAll(s, "İ", "i")
	s = strings.ReplaceAll(s, "I", "ı")
	return strings.ToLower(s)
}
func loadDictionary() {
	dictionaryWords = nil
	scan := bufio.NewScanner(strings.NewReader(dictionaryData))
	for scan.Scan() {
		line := strings.TrimSpace(scan.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		word := strings.Fields(line)[0]
		word = normalizeWord(strings.TrimSuffix(word, "'"))
		if wordPattern.MatchString(word) {
			dictionaryWords = append(dictionaryWords, word)
		}
	}
	sort.Strings(dictionaryWords)
}
func isDictionaryWord(word string) bool {
	dictionaryOnce.Do(loadDictionary)
	i := sort.SearchStrings(dictionaryWords, word)
	return i < len(dictionaryWords) && dictionaryWords[i] == word
}
func (b *Bot) handleGame(m *discordgo.MessageCreate) bool {
	s, e := b.db.Settings(m.GuildID)
	if e != nil {
		return false
	}
	counting := s.CountingEnabled && s.CountingChannelID.Valid && s.CountingChannelID.String == m.ChannelID
	word := s.WordChainEnabled && s.WordChainChannelID.Valid && s.WordChainChannelID.String == m.ChannelID
	if !counting && !word {
		return false
	}
	lockAny, _ := b.gameLocks.LoadOrStore(m.GuildID, &sync.Mutex{})
	lock := lockAny.(*sync.Mutex)
	lock.Lock()
	defer lock.Unlock()
	if counting {
		b.counting(m)
	} else {
		b.wordChain(m)
	}
	return true
}
func (b *Bot) rejectGame(m *discordgo.MessageCreate, reason string) {
	if b.db.ConfigBool(m.GuildID, "game_delete_invalid", true) {
		_ = b.dg.ChannelMessageDelete(m.ChannelID, m.ID)
	}
	sent, _ := b.dg.ChannelMessageSend(m.ChannelID, "❌ <@"+m.Author.ID+"> "+reason)
	if sent != nil {
		time.AfterFunc(5500*time.Millisecond, func() { _ = b.dg.ChannelMessageDelete(sent.ChannelID, sent.ID) })
	}
}
func (b *Bot) counting(m *discordgo.MessageCreate) {
	g, e := b.db.Game(m.GuildID)
	if e != nil {
		return
	}
	expected := g.CountingValue + 1
	content := strings.TrimSpace(m.Content)
	value, ok := parsePositive(content)
	same := g.CountingUserID.Valid && g.CountingUserID.String == m.Author.ID
	if !ok || value != expected || same {
		reset := b.db.ConfigBool(m.GuildID, "counting_reset_on_error", true)
		if reset {
			_ = b.db.ResetGame(m.GuildID, "counting")
		}
		if same {
			b.rejectGame(m, "aynı kişi art arda sayamaz."+str(reset, " Oyun **1**'den yeniden başladı.", ""))
		} else {
			b.rejectGame(m, "sıradaki sayı **"+itoa64(expected)+"** olmalıydı."+str(reset, " Oyun **1**'den yeniden başladı.", ""))
		}
		return
	}
	_ = b.db.SetCount(m.GuildID, expected, m.Author.ID)
	emoji := "✅"
	if expected%100 == 0 {
		emoji = "💯"
	}
	_ = b.dg.MessageReactionAdd(m.ChannelID, m.ID, emoji)
}
func parsePositive(s string) (int64, bool) {
	if !regexp.MustCompile(`^\d+$`).MatchString(s) {
		return 0, false
	}
	var n int64
	for _, r := range s {
		n = n*10 + int64(r-'0')
	}
	return n, true
}
func itoa64(v int64) string {
	if v == 0 {
		return "0"
	}
	var b [20]byte
	i := len(b)
	for v > 0 {
		i--
		b[i] = byte(v%10) + '0'
		v /= 10
	}
	return string(b[i:])
}
func (b *Bot) wordChain(m *discordgo.MessageCreate) {
	word := normalizeWord(m.Content)
	minLen := b.db.ConfigInt(m.GuildID, "word_min_length", 2)
	if len([]rune(word)) < minLen || !wordPattern.MatchString(word) {
		b.rejectGame(m, "yalnızca Türkçe tek bir kelime yazmalısın.")
		return
	}
	g, e := b.db.Game(m.GuildID)
	if e != nil {
		return
	}
	if g.WordUserID.Valid && g.WordUserID.String == m.Author.ID {
		b.rejectGame(m, "aynı kişi art arda kelime yazamaz.")
		return
	}
	if g.LastWord.Valid {
		last := []rune(g.LastWord.String)
		cur := []rune(word)
		if len(last) > 0 && cur[0] != last[len(last)-1] {
			b.rejectGame(m, "kelime **"+strings.ToUpper(string(last[len(last)-1]))+"** harfiyle başlamalı.")
			return
		}
	}
	if b.db.UsedWord(m.GuildID, word) {
		b.rejectGame(m, "**"+word+"** daha önce kullanıldı.")
		return
	}
	if !isDictionaryWord(word) {
		b.rejectGame(m, "**"+word+"** yerel Türkçe oyun sözlüğünde bulunamadı.")
		return
	}
	if b.db.SetWord(m.GuildID, word, m.Author.ID) == nil {
		_ = b.dg.MessageReactionAdd(m.ChannelID, m.ID, "✅")
	}
}
